File: tests/unit/engine/statistics.test.ts
import { describe, it, expect } from 'vitest';
import {
  calcCAGR,
  calcMWRR,
  calcAnnualizedStdev,
  calcSharpe,
  calcSortino,
  calcMaxDrawdown,
  calcCorrelation,
  calcDailyReturns,
} from '../../../packages/backend/src/engine/statistics.js';

describe('calcCAGR', () => {
  it('计算正常CAGR', () => {
    // 10000 → 20000，5年，CAGR ≈ 14.87%
    const cagr = calcCAGR(10000, 20000, 5);
    expect(cagr).toBeCloseTo(0.1487, 3);
  });

  it('零增长返回0', () => {
    expect(calcCAGR(10000, 10000, 5)).toBe(0);
  });

  it('负增长返回负值', () => {
    const cagr = calcCAGR(10000, 5000, 5);
    expect(cagr).toBeLessThan(0);
  });

  it('非法输入返回0', () => {
    expect(calcCAGR(0, 20000, 5)).toBe(0);
    expect(calcCAGR(10000, 0, 5)).toBe(0);
    expect(calcCAGR(10000, 20000, 0)).toBe(0);
    expect(calcCAGR(-10000, 20000, 5)).toBe(0);
  });
});

describe('calcMWRR', () => {
  it('简单投入场景', () => {
    // 投入10000，5年后收回20000
    const mwrr = calcMWRR([
      { value: -10000, time: 0 },
      { value: 20000, time: 5 },
    ]);
    expect(mwrr).toBeCloseTo(0.1487, 2);
  });

  it('空现金流返回0', () => {
    expect(calcMWRR([])).toBe(0);
  });
});

describe('calcAnnualizedStdev', () => {
  it('常数收益率波动率为0', () => {
    expect(calcAnnualizedStdev([0.01, 0.01, 0.01, 0.01])).toBe(0);
  });

  it('有波动的收益率返回正值', () => {
    const stdev = calcAnnualizedStdev([0.01, -0.02, 0.03, -0.01, 0.02]);
    expect(stdev).toBeGreaterThan(0);
  });
});

describe('calcSharpe', () => {
  it('正常计算', () => {
    const sharpe = calcSharpe(0.1, 0.15);
    // (0.10 - 0.02) / 0.15 ≈ 0.533
    expect(sharpe).toBeCloseTo(0.533, 2);
  });

  it('波动率为0返回0', () => {
    expect(calcSharpe(0.1, 0)).toBe(0);
  });
});

describe('calcSortino', () => {
  it('正常计算：混合正负收益', () => {
    // cagr=0.12, 日收益率含正负
    const sortino = calcSortino(0.12, [0.001, -0.002, 0.003, -0.001, 0.002]);
    expect(sortino).toBeGreaterThan(0);
  });

  it('全部收益为正且 cagr>无风险利率返回 99.9', () => {
    // 所有日收益率高于无风险日利率 → 下行波动为零
    const sortino = calcSortino(0.15, [0.001, 0.002, 0.003, 0.0015, 0.0025]);
    expect(sortino).toBe(99.9);
  });

  it('全部收益为正但 cagr<=无风险利率返回 0', () => {
    const sortino = calcSortino(0.01, [0.001, 0.002, 0.003]);
    expect(sortino).toBe(0);
  });

  it('空数组返回 0', () => {
    expect(calcSortino(0.12, [])).toBe(0);
  });

  it('单个元素返回 0', () => {
    expect(calcSortino(0.12, [0.001])).toBe(0);
  });
});

describe('calcMaxDrawdown', () => {
  it('单调递增无回撤', () => {
    const { maxDrawdown, maxDrawdownDuration } = calcMaxDrawdown([100, 110, 120, 130]);
    expect(maxDrawdown).toBe(0);
    expect(maxDrawdownDuration).toBe(0);
  });

  it('简单回撤', () => {
    // 100 → 120 → 90 → 110
    // 最大回撤 = (120-90)/120 = 25%
    const { maxDrawdown } = calcMaxDrawdown([100, 120, 90, 110]);
    expect(maxDrawdown).toBeCloseTo(0.25, 3);
  });
});

describe('calcCorrelation', () => {
  it('完全正相关', () => {
    const r = calcCorrelation([1, 2, 3, 4], [2, 4, 6, 8]);
    expect(r).toBeCloseTo(1, 5);
  });

  it('完全负相关', () => {
    const r = calcCorrelation([1, 2, 3, 4], [8, 6, 4, 2]);
    expect(r).toBeCloseTo(-1, 5);
  });
});

describe('数据不足时的边缘情况', () => {
  it.each([
    { name: 'calcAnnualizedStdev([0.01])', fn: () => calcAnnualizedStdev([0.01]), expected: 0 },
    { name: 'calcAnnualizedStdev([])', fn: () => calcAnnualizedStdev([]), expected: 0 },
    {
      name: 'calcMaxDrawdown([100])',
      fn: () => calcMaxDrawdown([100]),
      expected: { maxDrawdown: 0, maxDrawdownDuration: 0 },
    },
    { name: 'calcCorrelation([1], [2])', fn: () => calcCorrelation([1], [2]), expected: 0 },
  ])('$name', ({ fn, expected }) => {
    expect(fn()).toEqual(expected);
  });
});

describe('calcDailyReturns', () => {
  it('正常计算', () => {
    const returns = calcDailyReturns([100, 110, 105]);
    expect(returns).toHaveLength(2);
    expect(returns[0]).toBeCloseTo(0.1, 5); // (110-100)/100
    expect(returns[1]).toBeCloseTo(-0.04545, 4); // (105-110)/110
  });

  it('前值为0返回0', () => {
    expect(calcDailyReturns([0, 100])).toEqual([0]);
  });
});

File: engine-go/internal/engine/statistics_test.go
package engine

import (
	"engine-go/internal/enginetest"
	"engine-go/internal/mathutil"
	"math"
	"testing"
)

func TestCalcCAGR(t *testing.T) {
	for _, tc := range []struct {
		name                              string
		startValue, endValue, years, want float64
	}{
		{"doubles in 1 year", 100, 200, 1, 1.0},
		{"no growth", 100, 100, 1, 0},
		{"10% for 5 years", 100, 161.051, 5, 0.1},
		{"zero start", 0, 100, 1, 0},
		{"zero end", 100, 0, 1, 0},
		{"negative years", 100, 200, -1, 0},
		{"small values", 0.001, 0.002, 1, 1.0},
	} {
		t.Run(tc.name, func(t *testing.T) {
			assertFloatApprox(t, CalcCAGR(tc.startValue, tc.endValue, tc.years), tc.want, "CalcCAGR")
		})
	}
}
func TestCalcMWRR(t *testing.T) {
	for _, tc := range []struct {
		name      string
		cashflows []Cashflow
		want      float64
	}{
		{"no cashflows", nil, 0},
		{"invest 100 receive 120 in 1yr", []Cashflow{{-100, 0}, {120, 1}}, 0.2},
		{"invest 100 receive 110 in 1yr", []Cashflow{{-100, 0}, {110, 1}}, 0.1},
		{"only inflows undefined", []Cashflow{{100, 1}}, 0},
		{"only outflows undefined", []Cashflow{{-100, 0}}, 0},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if got := CalcMWRR(tc.cashflows); math.Abs(got-tc.want) > 1e-6 {
				t.Errorf("CalcMWRR() = %v, want %v", got, tc.want)
			}
		})
	}
}
func TestCalcDiversificationRatio(t *testing.T) {
	for _, tc := range []struct {
		name      string
		weights   []float64
		assets    [][]float64
		portfolio []float64
		want      float64
	}{
		{"mismatched lengths", []float64{1}, nil, []float64{0.01, -0.01, 0.02}, 0},
		{"short portfolio", []float64{1}, [][]float64{{0.01}}, []float64{0.01}, 0},
		{"zero portfolio stdev", []float64{1}, [][]float64{{0.01, -0.01}}, []float64{0, 0}, 0},
		{"single asset ratio 1", []float64{1}, [][]float64{{0.01, -0.01, 0.02}}, []float64{0.01, -0.01, 0.02}, 1},
	} {
		t.Run(tc.name, func(t *testing.T) {
			assertFloatApprox(t, CalcDiversificationRatio(tc.weights, tc.assets, tc.portfolio), tc.want, "CalcDiversificationRatio")
		})
	}
}
func TestMWRRWipeout(t *testing.T) {
	stats := CalculateStatisticsFromRequest(StatisticsRequest{
		Values:        []float64{100, 50, 0},
		Dates:         []string{"2024-01-01", "2024-06-01", "2025-01-01"},
		StartingValue: 100,
	})
	if stats.MWRR != -1 {
		t.Errorf("MWRR on total loss must be -1 (not 0), got %v", stats.MWRR)
	}
}
func TestCalcAnnualizedStdev(t *testing.T) {
	for _, tc := range []struct {
		name         string
		dailyReturns []float64
		want         float64
	}{
		{"empty", nil, 0},
		{"single return", []float64{0.01}, 0},
		{"two returns symmetric", []float64{0.01, -0.01}, math.Sqrt(0.0002) * math.Sqrt(252)},
		{"all zeros", []float64{0, 0, 0}, 0},
	} {
		t.Run(tc.name, func(t *testing.T) {
			assertFloatApprox(t, CalcAnnualizedStdev(tc.dailyReturns), tc.want, "CalcAnnualizedStdev")
		})
	}
}
func TestCalcRatioFuncs(t *testing.T) {
	for _, tc := range []struct {
		name string
		fn   func(float64, float64) float64
		a, b float64
		want float64
	}{
		{"Sharpe zero stdev", CalcSharpe, 0.10, 0, 0},
		{"Sharpe positive", CalcSharpe, 0.10, 0.15, (0.10 - 0.02) / 0.15},
		{"Sharpe negative cagr", CalcSharpe, -0.05, 0.20, (-0.05 - 0.02) / 0.20},
		{"Calmar zero drawdown", CalcCalmar, 0.10, 0, 0},
		{"Calmar normal case", CalcCalmar, 0.10, 0.20, 0.5},
		{"Calmar negative cagr", CalcCalmar, -0.05, 0.20, -0.25},
	} {
		t.Run(tc.name, func(t *testing.T) { assertFloatApprox(t, tc.fn(tc.a, tc.b), tc.want, "ratio") })
	}
}

// U-2 收尾：CalcUPI 改显式 rf（与 MartinRatio 口径一致）；rf=0.02 时与旧实现数值相同
func TestCalcUPI(t *testing.T) {
	for _, tc := range []struct {
		name               string
		cagr, ulcerIdx, rf float64
		want               float64
	}{
		{"legacy rf 0.02 same as before", 0.10, 0.15, 0.02, (0.10 - 0.02) / 0.15},
		{"zero ulcer", 0.10, 0, 0.02, 0},
		{"explicit non-legacy rf", 0.10, 0.15, 0.05, (0.10 - 0.05) / 0.15},
		{"negative cagr", -0.05, 0.20, 0.02, (-0.05 - 0.02) / 0.20},
	} {
		t.Run(tc.name, func(t *testing.T) {
			assertFloatApprox(t, CalcUPI(tc.cagr, tc.ulcerIdx, tc.rf), tc.want, "CalcUPI")
		})
	}
}
func TestCalcSortino(t *testing.T) {
	for _, tc := range []struct {
		name     string
		cagr     float64
		returns  []float64
		wantZero bool
	}{
		{"insufficient data", 0.10, []float64{0.01}, true},
		{"no downside returns above daily risk free", 0.10, []float64{0.01, 0.02, 0.03}, true},
		{"all downside returns negative", 0.10, []float64{-0.01, -0.02}, false},
	} {
		t.Run(tc.name, func(t *testing.T) {
			got := CalcSortino(tc.cagr, tc.returns)
			if tc.wantZero {
				if got != 0 {
					t.Errorf("CalcSortino() = %v, want 0", got)
				}
			} else {
				if got == 0 {
					t.Error("CalcSortino() = 0, expected non-zero for negative returns")
				}
			}
		})
	}
}
func TestCalcMaxDrawdown(t *testing.T) {
	for _, tc := range []struct {
		name         string
		values       []float64
		wantDrawdown float64
		wantDuration int
	}{
		{"insufficient data", []float64{100}, 0, 0},
		{"monotonic up", []float64{100, 110, 120, 130}, 0, 0},
		{"monotonic down", []float64{100, 90, 80, 70}, 0.3, 3},
		{"peak recovery", []float64{100, 110, 90, 80, 110}, (110.0 - 80.0) / 110.0, 2},
		{"two peaks higher recovery", []float64{100, 110, 105, 115, 105, 95}, (115.0 - 95.0) / 115.0, 2},
	} {
		t.Run(tc.name, func(t *testing.T) {
			got := CalcMaxDrawdown(tc.values)
			if math.Abs(got.MaxDrawdown-tc.wantDrawdown) > 1e-10 {
				t.Errorf("MaxDrawdown = %v, want %v", got.MaxDrawdown, tc.wantDrawdown)
			}
			if got.MaxDrawdownDuration != tc.wantDuration {
				t.Errorf("MaxDrawdownDuration = %v, want %v", got.MaxDrawdownDuration, tc.wantDuration)
			}
		})
	}
}
func TestCalcCorrelation(t *testing.T) {
	for _, tc := range []struct {
		name string
		a, b []float64
		want float64
	}{
		{"insufficient data", []float64{1}, []float64{2}, 0},
		{"perfect correlation", []float64{1, 2, 3}, []float64{2, 4, 6}, 1},
		{"inverse correlation", []float64{1, 2, 3}, []float64{3, 2, 1}, -1},
		{"no correlation (constant)", []float64{1, 2, 3}, []float64{1, 1, 1}, 0},
		{"truncated to shorter", []float64{1, 2, 3, 4}, []float64{2, 4, 6}, 1},
	} {
		t.Run(tc.name, func(t *testing.T) { assertFloatApprox(t, CalcCorrelation(tc.a, tc.b), tc.want, "CalcCorrelation") })
	}
}
func TestCalcDailyReturns(t *testing.T) {
	for _, tc := range []struct {
		name   string
		prices []float64
		want   []float64
	}{
		{"insufficient data", []float64{100}, nil},
		{"two prices", []float64{100, 110}, []float64{0.1}},
		{"three prices", []float64{100, 110, 121}, []float64{0.1, 0.1}},
		{"declining", []float64{100, 90, 81}, []float64{-0.1, -0.1}},
		{"缺失缺口应按无变动", []float64{100, 0, 110}, []float64{0}},
		{"真实清零应记 -100%", []float64{100, 0, 0}, []float64{-1}},
		{"缺口后恢复应正常", []float64{100, 0, 110, 121}, []float64{0, 0.1}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			got := mathutil.DailyReturns(tc.prices)
			if len(got) != len(tc.want) {
				t.Fatalf("len = %v, want %v", len(got), len(tc.want))
			}
			for i := range got {
				assertFloatApprox(t, got[i], tc.want[i], "DailyReturns[i]")
			}
		})
	}
}
func TestCalcDailyReturnsWithZeros(t *testing.T) {
	for _, tc := range []struct {
		name   string
		prices []float64
		want   []float64
	}{
		{"insufficient data", []float64{100}, nil},
		{"regular", []float64{100, 110, 121}, []float64{0.1, 0.1}},
		{"缺失缺口应按无变动", []float64{100, 0, 110}, []float64{0, 0}},
		{"真实清零应记 -100%", []float64{100, 0, 0}, []float64{-1, 0}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			got := mathutil.DailyReturnsWithZeros(tc.prices)
			if len(got) != len(tc.want) {
				t.Fatalf("len = %v, want %v", len(got), len(tc.want))
			}
			for i := range got {
				assertFloatApprox(t, got[i], tc.want[i], "DailyReturnsWithZeros[i]")
			}
		})
	}
}
func TestCalcTotalReturn(t *testing.T) {
	for _, tc := range []struct {
		name       string
		start, end float64
		want       float64
	}{
		{"zero start", 0, 100, 0},
		{"negative start", -100, 200, 0},
		{"100% gain", 100, 200, 1.0},
		{"50% loss", 100, 50, -0.5},
		{"no change", 100, 100, 0},
	} {
		t.Run(tc.name, func(t *testing.T) {
			assertFloatApprox(t, CalcTotalReturn(tc.start, tc.end), tc.want, "CalcTotalReturn")
		})
	}
}
func TestMinMaxValue(t *testing.T) {
	for _, tc := range []struct {
		name string
		fn   func([]float64) float64
		vals []float64
		want float64
	}{
		{"MaxValue empty", MaxValue, nil, 0},
		{"MaxValue all positive", MaxValue, []float64{0.05, 0.10, 0.15}, 0.15},
		{"MaxValue mixed", MaxValue, []float64{-0.10, 0.20, -0.05}, 0.20},
		{"MinValue empty", MinValue, nil, 0},
		{"MinValue mixed", MinValue, []float64{0.05, -0.10, 0.15}, -0.10},
		{"MinValue all negative", MinValue, []float64{-0.05, -0.10, -0.15}, -0.15},
	} {
		t.Run(tc.name, func(t *testing.T) { assertFloatApprox(t, tc.fn(tc.vals), tc.want, "value") })
	}
}
func TestCalcDrawdownMetrics(t *testing.T) {
	for _, tc := range []struct {
		name string
		fn   func([]float64) float64
		vals []float64
		want float64
	}{
		{"AvgDrawdown insufficient data", CalcAvgDrawdown, []float64{100}, 0},
		{"AvgDrawdown monotonic up", CalcAvgDrawdown, []float64{100, 110, 120}, 0},
		{"AvgDrawdown single drawdown", CalcAvgDrawdown, []float64{100, 110, 90, 80}, ((110.0-90.0)/110.0 + (110.0-80.0)/110.0) / 2.0},
		{"UlcerIndex insufficient data", CalcUlcerIndex, []float64{100}, 0},
		{"UlcerIndex monotonic up", CalcUlcerIndex, []float64{100, 110, 120}, 0},
		{"UlcerIndex monotonic down", CalcUlcerIndex, []float64{100, 80, 60}, math.Sqrt(((100.0-80.0)*(100.0-80.0)/(100.0*100.0) + (100.0-60.0)*(100.0-60.0)/(100.0*100.0)) / 3.0)},
	} {
		t.Run(tc.name, func(t *testing.T) { assertFloatApprox(t, tc.fn(tc.vals), tc.want, "metric") })
	}
}
func swrVolatileRequest() StatisticsRequest {
	return StatisticsRequest{Values: []float64{100, 110}, AnnualReturnValues: enginetest.VolatileAnnualReturns(50)}
}
func TestSWRNotEqualToPWR(t *testing.T) {
	stats := CalculateStatisticsFromRequest(swrVolatileRequest())
	if stats.SWR == stats.PWR {
		t.Errorf("SWR (%v) should differ from PWR (%v) for volatile series", stats.SWR, stats.PWR)
	}
	if stats.SWR < 0 {
		t.Errorf("SWR should be non-negative, got %v", stats.SWR)
	}
}
func TestSWRUsesLongestStandardTerm(t *testing.T) {
	stats := CalculateStatisticsFromRequest(swrVolatileRequest())
	if stats.SWR != stats.SWR40Y {
		t.Errorf("SWR (%v) should equal SWR40Y (%v) when >=40 years of data available", stats.SWR, stats.SWR40Y)
	}
	if stats.SWR > stats.SWR30Y {
		t.Errorf("SWR (%v) should be <= SWR30Y (%v) (longer term = more conservative)", stats.SWR, stats.SWR30Y)
	}
}
